### ISSUE TITLE
NVDA considers words differently than caret navigation in Java application

### ISSUE THREAD

**Issue by raducoravu** [nvaccess/nvda#11183]
Let's say I have a Java text area with some content inside:

    public static void main(String[] args) {
      JFrame fr = new JFrame();
      fr.setSize(500, 500);
      JTextArea jt = new JTextArea("    <body id=\"abc_def_hij\">");
      fr.getContentPane().add(jt);
      fr.setVisible(true);
    }

I place the cursor before the "id" string, NVDA reads to me the entire text:

   id="abc_def_hij"

, but if I use Ctrl-Arrow Right navigation in the text area, it skips only over the "id" string.

In the Java accessibility support there are methods like:

javax.accessibility.AccessibleText.getAtIndex(int, int)

    /**
     * Returns the String at a given index.
     *
     * @param part the CHARACTER, WORD, or SENTENCE to retrieve
     * @param index an index within the text
     * @return the letter, word, or sentence
     */
    public String getAtIndex(int part, int index);

That "part" can also be of type WORD in order to ask the text component to return a word, but NVDA only asks for characters and makes its own decisions about word boundaries, decisions which may be different than the text area word boundary computation.

**Comment by leonardder** [comment]
Actually, we do not use `getAtIndex` at all. This is because our offsets based implementation works differently. When you use word navigation, NVDA:

1. gets the current caret offset
2. calls getWordOffsets with the caret offsets to get the word offsets
3. calls getTextRange with these offsets to get the actual word. 

This API expects us to skip the second step. However, this step is pretty essential.

**Comment by raducoravu** [comment]
Not sure what the solution could be, the Java swing JTextArea has in my opinion a better way to decide word boundaries plus various JTextArea custom implementations might have their own way of deciding what a word boundary is. 
Ideally NVDA would use the API to let the component decide what the word boundaries are.
Second best would be if NVDA would work on improving its word range computation, a sequence like this:

    id="abc_def_hij"

is for now considered one word by NVDA although maybe it should stop at characters which are not alphanumeric.

**Comment by lukaszgo1** [comment]
For what its worth JAWS respects Java word boundaries in this case.

**Comment by raducoravu** [comment]
Both the Java swing next word navigation method (javax.swing.text.Utilities.getNextWordInParagraph(JTextComponent, Element, int, boolean)) and the accessibility API implementation (javax.swing.text.JTextComponent.AccessibleJTextComponent.getSegmentAt(int, int)) are using the Java word break iterator "java.text.BreakIterator.getWordInstance()".

I created a small Java test using the word break iterator:

    public static void main(String[] args) {
    BreakIterator iter = BreakIterator.getWordInstance();
    String txt = "<a b=\"c_d\"/>";
    iter.setText(txt);
    
    int nxt = -1;
    int current = 0;
    while((nxt = iter.next()) != BreakIterator.DONE) {
      System.err.println(txt.substring(current, nxt));
      current = nxt;
       }
      }

and breaks the content like this:

      <
      a
       
      b
      =
      "
      c_d
      "
      /
      >

From what I seemed to remember in the tests I made a couple of days ago NVDA seemed to break the same content in larger chunks.

**Comment by Adriani90** [comment]
cc: @michaelDCurran

**Comment by jcsteh** [comment]
Looking briefly, JAB has a getAccessibleTextLineBounds function, but I don't see an equivalent for line. There is this GetAccessibleTextItems function which returns (I think?) an AccessibleTextItemsInfo pointer, and AccessibleTextItemsInfo has a "word" member, but I don't really understand what it does and I can't find any documentation. In typical JAB fashion, it's a total, unmaintained mess.

**Comment by jcsteh** [comment]
Okay, so with the Python console, you can, for example:
nav.jabContext.getAccessibleTextItems(8).word
(where 8 is the offset you want the word for)
That'll give you the text of the word at that offset, but that's not sufficient for NVDA. NVDA needs the offsets of the word, not just the text. Otherwise, it wouldn't be able to manage the review cursor, etc.

I guess one possibility might be to do a `_getWordOffsets` implementation which reimplements the Java word break algorithm, whatever that is specifically. That wouldn't work if someone did a custom a11y implementation, though.

**Comment by jcsteh** [comment]
Or you could have it fetch the word text using getAccessibleTextItems, get the text from (offset - wordLength, offset + wordLength) and search that text for an instance of the word which encompasses the offset. Then, you'd have the start and end offsets. But seriously, yuck. Sometimes, I horrify myself.

**Comment by raducoravu** [comment]
It's a pity the "AccessibleTextItemsInfo" does not contain the interval offsets for the word. So indeed there would be a need to use it in conjunction with other methods to obtain the interval offsets for the current word.
The Java implementation of the word break iterator is probably dependent on the locale, also I cannot find an implementation for it, it's probably in the not exposed sun.* packages. 
With the current way in which this works, when the cursor is placed before this text:

    id="abc_def_hij"

NVDA reads "id="abc_def_hij"", then the end user uses Ctrl-Right arrow (which delegates to the JTextArea next word actions) to jump over the word, and this action jumps only over the "id" attribute, so NVDA reads again "="abc_def_hij"". So in such cases the end user hears a lot of repeated content when using Ctrl-Right arrow to navigate the content word by word.

**Comment by Adriani90** [comment]
cc: @mwhapples, @dmitrii-drobotov, @raducoravu, @aaryan-kumar  I guess this issue is still the same in current versions of JAB and NVDA 2024.4.1 right?

@raducoravu wrote:
> Second best would be if NVDA would work on improving its word range computation, a sequence like this:
id="abc_def_hij"
is for now considered one word by NVDA 

Actually I still think this is the first best option, because now we see lots of different word handling in different applications which makes word navigation very inconsistent. Having a specific computation in NVDA of a word, considering also mathematical contexts and other languages which do not use space bar to split sentences, is the best way to go in my view. But I know at the same time that this opinion is seen as controversial within the community though,. Just wanted to mention it.
Actually NVDA can already make the difference between alphanumeric and non alphanumeric characters, but this is not applied consistently in the word navigation. I hope software developers join forces and submit a proposal to the NVDA repository for a word computation that makes sense from an accessibility point of view, and so we can get the community involved.

> although maybe it should stop at characters which are not alphanumeric.

This is not that simple, non latin script languages such as chinese group languages do not use alphabethical characters as far as I understand. Still they need word by word navigation as well.

cc: @cary-rowen

**Comment by dmitrii-drobotov** [comment]
@Adriani90 yes, this issue is still relevant, and I'm actually already working on it in the scope of #16374, I wasn't aware of this one. It is a tricky issue because apps like IntelliJ allow configuring the way the caret moves with Ctrl+Left/Right, and I think it would be better for NVDA to read words the way they are configured. For example, a string `someLongVariableName` can be read as a single word or as 4 separate words depending on the settings.

**Comment by Adriani90** [comment]
I see. For paragraph navigation at least, we can decide in NVDA settings whether caret should handled by application, or by NVDA‘s own paragraph definition. This optionality improved lots of use cases. Von meinem iPhone gesendetAm 21.11.2024 um 13:28 schrieb Dmitrii Drobotov ***@***.***>:﻿
@Adriani90 yes, this issue is still relevant, and I'm actually already working on it in the scope of #16374, I wasn't aware of this one. It is a tricky issue because apps like IntelliJ allow configuring the way the caret moves with Ctrl+Left/Right, and I think it would be better for NVDA to read words the way they are configured. For example, a string someLongVariableName can be read as a single word or as 4 separate words depending on the settings.

**Comment by CyrilleB79** [comment]
Yes, the "Paragraph style" option in "Document Navigation" settings allows a fine tuning of the way NVDA navigates with control+up/downArrow. A similar option would be welcome for word navigation, i.e. doing more or less what @mltony's WordNav add-on does.

NVDA would not automatically guess how the application implements word navigation, especially if it is configurable in the application (such as IntelliJ). But at least, implementing such option would allow the user to define a specific application profile where the word navigation setting in NVDA would be defined to something equivalent to what the user has defined in the application.

**Comment by mltony** [comment]
Yes we proposed to include wordNav into NVDA core in #16237 but nvAccess weren't aligned, at least on the general idea. Specifically Sean mentioned that they don't consider adding developer-centric functionality to the core.
I personally still think that including wordNav in the core might be still a good idea precisely because it fundamentally gets rid of this discrepancy problem where NVDA word definition is different from application word definition.

**Comment by CyrilleB79** [comment]
@mltony see https://github.com/nvaccess/nvda/issues/16237#issuecomment-2492392480.

**Comment by kranid** [comment]
@CyrilleB79 NVDA can automatically determine how the application implements word navigation by utilizing the corresponding Java Accessibility API, and this capability is particularly important when the word navigation method is configurable. Even if the Word Nav feature will be integrated into NVDA Core, NVDA should provide the option to use the word navigation mode defined within the application, and I believe this should be the default option.

**Comment by CyrilleB79** [comment]
> Even if the Word Nav feature will be integrated into NVDA Core, NVDA should provide the option to use the word navigation mode defined within the application, and I believe this should be the default option.

Yes I fully agree.

**Comment by jcsteh** [comment]
> NVDA can automatically determine how the application implements word navigation by utilizing the corresponding Java Accessibility API

It can't, though. The API provided by Java Access Bridge (which is how external clients like NVDA access Java Swing apps) isn't sufficient. See https://github.com/nvaccess/nvda/issues/11183#issuecomment-634000174. NVDA needs offsets, not just text.

**Comment by kranid** [comment]
> NVDA needs offsets, not just text.

Yes, but NVDA can calculate offsets if it gets the word.
yeas, but NVDA can caltulate offsets if it gets  the word.

**Comment by jcsteh** [comment]
It possibly can't do even that reliably. See https://github.com/nvaccess/nvda/issues/11183#issuecomment-634007190. Note that this also gets a lot messier when dealing with 16 or 32 bit Unicode characters depending on the various encoding/decoding that occurs throughout the stack.

But sure, if someone wants to try implementing this, https://github.com/nvaccess/nvda/issues/11183#issuecomment-634007190 is where I would start trying.

**Comment by dmitrii-drobotov** [comment]
@jcsteh Yes, I actually already implemented it this way in my local branch, and it seems to work fine. Here's the code if you're interested: https://github.com/nvaccess/nvda/commit/eb756057aaa7ad85653bb1c24bdad1e68099717e. Thanks for pointing out about possible issues with unicode characters, I will do more testing with it. But from the first look, it seems to handle it well, I tried with Chinese characters and emojis. In any case, if it fails to find the word, it will fall back to the current implementation. I understand that this is an inefficient approach and may not handle some edge cases, but I don't see a better way.

Also, to clarify, I haven't submitted the PR with this fix yet because I've found another issue on the IntelliJ side. While the code editor component provides correct word bounds information depending on the settings, other text fields return the same word info regardless of the settings. Basically, the caret movement settings affect almost all text fields in the app, but the accessibility info is adjusted only for the editor, so in those other text fields it might become worse. That's why I would like to fix this issue in IntelliJ first.

@mltony I checked the WordNav add-on with IntelliJ and it works well out of the box. I also really like the idea of this approach, great work! I did find one related issue that was fixed somewhere in the IntelliJ 2023.1 release, so it might not work with earlier versions. If you get other issues reported with IntelliJ and your add-ons, feel free to forward them to me and I can look into it.

**Comment by kranid** [comment]
@dmitrii-drobotov When I use Word nav with IDEA, I face a problem when IDEA cannot identify the correct caret position. This problem affects caret position-dependent operations, such as 'Go to Implementation' or 'Go to Definition'.".

**Comment by dmitrii-drobotov** [comment]
@kranid Interesting, for me using Ctrl+Right/Left arrows or Ctrl+Win+Right/Left with wordnav moves the actual caret in IntelliJ, so these actions work correctly. Did you maybe configure wordnav in some way?
